File: benchmarks/structbench/structbench.py
# structbench.py — Python reference for struct/enum benchmark.
#   python structbench.py [n]
import sys
import time


class Point:
    __slots__ = ("x", "y")

    def __init__(self, x, y):
        self.x = x
        self.y = y

    def dist2(self):
        return self.x * self.x + self.y * self.y


class Circle:
    __slots__ = ("center", "r")

    def __init__(self, center, r):
        self.center = center
        self.r = r

    def area(self):
        return 3.14159 * self.r * self.r


# tagged tuple, mirrors LS enum (tag, a, b)
def shape_measure(s):
    tag = s[0]
    if tag == 0:
        return 0.0
    elif tag == 1:
        return s[1]
    else:
        return s[1] * s[2]


def bench_scalar(n):
    s = 0.0
    for i in range(n):
        c = float(i % 1000)
        p = Point(c, c * 2.0)
        s += p.dist2()
    return s


def bench_nested(n):
    s = 0.0
    for i in range(n):
        c = float(i % 1000)
        cir = Circle(Point(c, 0.0), c)
        s += cir.area() + cir.center.x
    return s


def bench_enum(n):
    s = 0.0
    for i in range(n):
        c = float(i % 1000)
        k = i % 3
        sh = (0, 0.0, 0.0)
        if k == 1:
            sh = (1, c, 0.0)
        if k == 2:
            sh = (2, 2.0, c)
        s += shape_measure(sh)
    return s


def main():
    n = int(sys.argv[1]) if len(sys.argv) >= 2 else 1000000

    _w = bench_scalar(1000) + bench_nested(1000) + bench_enum(1000)

    t0 = time.perf_counter_ns()
    r1 = bench_scalar(n)
    t1 = time.perf_counter_ns()
    r2 = bench_nested(n)
    t2 = time.perf_counter_ns()
    r3 = bench_enum(n)
    t3 = time.perf_counter_ns()

    chk = int(r1 + r2 + r3)
    print(f"result: {chk}")
    print(f"[@bench] scalar    {(t1 - t0) // 1000} us")
    print(f"[@bench] nested    {(t2 - t1) // 1000} us")
    print(f"[@bench] enum      {(t3 - t2) // 1000} us")
    print(f"[@bench] mean {(t3 - t0) // 1000} us ({n} iterations)")


main()

File: benchmarks/structbench/structbench.cpp
// structbench.cpp — C++ (MSVC) reference for struct/enum benchmark.
//   cl /O2 /std:c++17 /EHsc /MT structbench.cpp /Fe:structbench_cpp.exe
#include <cstdio>
#include <cstdlib>
#include <cstdint>
#include <chrono>

struct Point {
    double x, y;
    double dist2() const { return x * x + y * y; }
};
struct Circle {
    Point center;
    double r;
    double area() const { return 3.14159 * r * r; }
};

// tagged union, mirrors LS enum Shape
struct Shape {
    int tag; // 0=Dot 1=Seg 2=Box
    double a, b;
};
static double shape_measure(const Shape &s) {
    switch (s.tag) {
        case 0: return 0.0;
        case 1: return s.a;
        case 2: return s.a * s.b;
    }
    return 0.0;
}

// prevent the optimizer from eliding the loops
static volatile double sink;

static double bench_scalar(long long n) {
    double s = 0.0;
    for (long long i = 0; i < n; i++) {
        double c = (double)(i % 1000);
        Point p { c, c * 2.0 };
        s += p.dist2();
    }
    return s;
}
static double bench_nested(long long n) {
    double s = 0.0;
    for (long long i = 0; i < n; i++) {
        double c = (double)(i % 1000);
        Circle cir { Point { c, 0.0 }, c };
        s += cir.area() + cir.center.x;
    }
    return s;
}
static double bench_enum(long long n) {
    double s = 0.0;
    for (long long i = 0; i < n; i++) {
        double c = (double)(i % 1000);
        int k = (int)(i % 3);
        Shape sh { 0, 0, 0 };
        if (k == 1) sh = Shape { 1, c, 0 };
        if (k == 2) sh = Shape { 2, 2.0, c };
        s += shape_measure(sh);
    }
    return s;
}

int main(int argc, char **argv) {
    long long n = (argc >= 2) ? atoll(argv[1]) : 1000000;
    sink = bench_scalar(1000) + bench_nested(1000) + bench_enum(1000);

    auto t0 = std::chrono::high_resolution_clock::now();
    double r1 = bench_scalar(n); sink = r1;
    auto t1 = std::chrono::high_resolution_clock::now();
    double r2 = bench_nested(n); sink = r2;
    auto t2 = std::chrono::high_resolution_clock::now();
    double r3 = bench_enum(n); sink = r3;
    auto t3 = std::chrono::high_resolution_clock::now();

    auto us = [](auto a, auto b) {
        return (long long)std::chrono::duration_cast<std::chrono::microseconds>(b - a).count();
    };
    long long chk = (long long)(r1 + r2 + r3);
    printf("result: %lld\n", chk);
    printf("[@bench] scalar    %lld us\n", us(t0, t1));
    printf("[@bench] nested    %lld us\n", us(t1, t2));
    printf("[@bench] enum      %lld us\n", us(t2, t3));
    printf("[@bench] mean %lld us (%lld iterations)\n", us(t0, t3), n);
    return 0;
}
